<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="link_0">

<!-- GEOMETRY 1 -->
    <xacro:macro name="link_0_geometry1">
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <geometry>
            <cylinder length="${link_0_length1}" radius="${link_0_radius1}"/>
        </geometry>
    </xacro:macro>

<!-- GEOMETRY 2 -->
    <xacro:macro name="link_0_geometry2">
        <origin rpy="1.57075 0 0" xyz="0 0 ${link_0_length1/2}"/>
        <geometry>
            <cylinder length="${link_0_length2}" radius="${link_0_radius2}"/>
        </geometry>
    </xacro:macro>

<!-- MATERIAL -->
    <xacro:macro name="link_0_material">
        <material name="blue">
            <color rgba=".2 .2 .5 1"/>
        </material>
    </xacro:macro>

<!-- LINK -->
    <xacro:macro name="link_0" params="name parent x y z">

<!-- LINK -->
        <link name="${name}">
            <visual>
                <xacro:link_0_geometry1/>
                <xacro:link_0_material/>
            </visual>
            <visual>
                <xacro:link_0_geometry2/>
                <xacro:link_0_material/>
            </visual>

            <collision>
                <xacro:link_0_geometry1/>
            </collision>
            <collision>
                <xacro:link_0_geometry2/>
            </collision>
            <inertial>
                 <mass value="1"/>
                 <inertia ixx=".003" ixy=".00" ixz=".00" iyy=".003" iyz=".00" izz=".003"/>
            </inertial>
        </link>
        <gazebo reference="${name}">
            <mu1> .2 </mu1>
            <mu2> .2 </mu2>
            <material> Gazebo/Grey </material>
            <selfCollide> False </selfCollide>
            <implicitSpringDamper> True </implicitSpringDamper>
        </gazebo>

<!-- JOINT -->
        <joint name="${name}_joint" type="revolute">
            <parent link="${parent}"/>
            <child link="${name}"/>
            <origin rpy="0 0 0" xyz="${x} ${y} ${z}"/>
            <axis xyz="0 0 -1"/>
            <limit effort="100" velocity="1" lower="-3" upper="3"/>
            <dynamics damping="1"/>
        </joint>
        <!--gazebo reference="${name}_joint">
            <stopCfm>0.8</stopCfm>
            <stopErp>0.5</stopErp>
            <provideFeedback>true</provideFeedback>
            <implicitSpringDamper>true</implicitSpringDamper>
            <fudgeFactor>0.0</fudgeFactor>
        </gazebo-->
        <gazebo>
            <plugin name="joint_limiter_${name}" filename="libJointLimiter.so">
                <joint>${name}_joint</joint>
                <topic>/simulation/controller/vel/${name}</topic>
                <lower>-3</lower>
                <upper>3</upper>
            </plugin>
        </gazebo>
        <gazebo>
            <plugin name="joint_encoder_${name}" filename="libJointEncoder.so">
                <joint>${name}_joint</joint>
                <topic>/simulation/encoder/abs/${name}</topic>
                <lower>-3</lower>
                <upper>3</upper>
                <absolute>1</absolute>
                <resolution>3600</resolution>
            </plugin>
        </gazebo>
        <gazebo>
            <plugin name="joint_controller_${name}" filename="libControllerPID.so">
                <joint>${name}_joint</joint>
                <topic>/simulation/controller/vel/${name}</topic>
                <Kp>20</Kp>
                <Ki>2</Ki>
                <Kd>1</Kd>
            </plugin>
        </gazebo>



<!-- TRANSMISSION -->
        <!--transmission name="${name}_transmission">
            <type>transmission_interface/SimpleTransmission</type>
            <joint name="${name}_joint">
                <hardwareInterface>EffortJointInterface</hardwareInterface>
            </joint>
            <actuator name="${name}_motor">
                <hardwareInterface>EffortJointInterface</hardwareInterface>
                <mechanicalReduction>1</mechanicalReduction>
            </actuator>
        </transmission-->

    </xacro:macro>
</robot>
